kernel/src/cpu/apic.rs

```diff
@@ -0,0 +1,851 @@
+        &mut self,
+        cpu_state: &mut T,
+        caa_addr: Option<VirtAddr>,
+    ) {
+        // Check to see if a previously delivered interrupt is still pending.
+        // If so, move it back to the IRR.
+        if self.interrupt_delivered {
+            let irq = cpu_state.check_and_clear_pending_interrupt_event();
+            if irq != 0 {
+                self.rewind_pending_interrupt(irq);
+                self.lazy_eoi_pending = false;
+            }
+            self.interrupt_delivered = false;
+        }
+
+        // Check to see if a previously queued interrupt is still pending.
+        // If so, move it back to the IRR.
+        if self.interrupt_queued {
+            let irq = cpu_state.check_and_clear_pending_virtual_interrupt();
+            if irq != 0 {
+                self.rewind_pending_interrupt(irq);
+                self.lazy_eoi_pending = false;
+            }
+            self.interrupt_queued = false;
+        }
+
+        // If a lazy EOI is pending, then check to see whether an EOI has been
+        // requested by the guest.  Note that if a lazy EOI was dismissed
+        // above, the guest lazy EOI flag need not be cleared here, since
+        // dismissal of any interrupt above will require reprocessing of
+        // interrupt state prior to guest reentry, and that reprocessing will
+        // reset the guest lazy EOI flag.
+        if self.lazy_eoi_pending {
+            if let Some(virt_addr) = caa_addr {
+                let calling_area = GuestPtr::<SvsmCaa>::new(virt_addr);
+                if let Ok(caa) = calling_area.read() {
+                    if caa.no_eoi_required == 0 {
+                        assert!(self.isr_stack_index != 0);
+                        self.perform_eoi();
+                    }
+                }
+            }
+        }
+    }
+
+    fn get_ppr_with_tpr(&self, tpr: u8) -> u8 {
+        // Determine the priority of the current in-service interrupt, if any.
+        let ppr = if self.isr_stack_index != 0 {
+            self.isr_stack[self.isr_stack_index]
+        } else {
+            0
+        };
+
+        // The PPR is the higher of the in-service interrupt priority and the
+        // task priority.
+        if (ppr >> 4) > (tpr >> 4) {
+            ppr
+        } else {
+            tpr
+        }
+    }
+
+    fn get_ppr<T: GuestCpuState>(&self, cpu_state: &T) -> u8 {
+        self.get_ppr_with_tpr(cpu_state.get_tpr())
+    }
+
+    fn clear_guest_eoi_pending(caa_addr: Option<VirtAddr>) -> Option<GuestPtr<SvsmCaa>> {
+        if let Some(virt_addr) = caa_addr {
+            let calling_area = GuestPtr::<SvsmCaa>::new(virt_addr);
+            // Ignore errors here, since nothing can be done if an error
+            // occurs.
+            if let Ok(caa) = calling_area.read() {
+                let _ = calling_area.write(caa.update_no_eoi_required(0));
+            }
+            Some(calling_area)
+        } else {
+            None
+        }
```

Comment:
If `caa_addr` is None you can just return early with `?`:

```suggestion
        let virt_addr = caa_addr?;
        let calling_area = GuestPtr::<SvsmCaa>::new(virt_addr);
        // Ignore errors here, since nothing can be done if an error occurs.
        if let Ok(caa) = calling_area.read() {
            let _ = calling_area.write(caa.update_no_eoi_required(0));
        }
        Some(calling_area)
```

Reply:
This is a clear improvement.